File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/account/application/query/GetAccountService.java
package com.example.accountservice.account.application.query;

import com.example.accountservice.account.domain.Account;
import com.example.accountservice.account.domain.AccountException;
import com.example.accountservice.account.domain.AccountFindQuery;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class GetAccountService {

    private final AccountQuery accountQuery;

    public GetAccountResult getAccount(String accountId, String requesterId) {
        Account account =
                accountQuery
                        .findAccounts(new AccountFindQuery(0, 1, accountId, requesterId, null))
                        .accounts()
                        .stream()
                        .findFirst()
                        .orElseThrow(
                                () ->
                                        new AccountException(
                                                AccountException.ErrorCode.ACCOUNT_NOT_FOUND,
                                                "Account not found."));
        return new GetAccountResult(
                account.getAccountId(),
                account.getOwnerId(),
                account.getEmail(),
                new GetAccountResult.MoneyResult(
                        account.getBalance().amount(), account.getBalance().currency()),
                account.getStatus().name(),
                account.getCreatedAt(),
                account.getUpdatedAt());
    }
}

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/account/application/query/GetAccountService.kt
package com.example.accountservice.account.application.query

import com.example.accountservice.account.domain.AccountFindQuery
import com.example.accountservice.account.domain.AccountNotFoundException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class GetAccountService(
    private val accountQuery: AccountQuery,
) {
    fun getAccount(
        accountId: String,
        requesterId: String,
    ): GetAccountResult {
        val (accounts, _) =
            accountQuery.findAccounts(
                AccountFindQuery(page = 0, take = 1, accountId = accountId, ownerId = requesterId),
            )
        val account = accounts.firstOrNull() ?: throw AccountNotFoundException(accountId)
        return GetAccountResult(
            accountId = account.accountId,
            ownerId = account.ownerId,
            email = account.email,
            balance = GetAccountResult.MoneyResult(account.balance.amount, account.balance.currency),
            status = account.status.name,
            createdAt = account.createdAt,
            updatedAt = account.updatedAt,
        )
    }
}
